lib/block.rb

```diff
@@ -0,0 +1,118 @@
+module HotelBooking
+  class Block
+    
+    attr_reader :id, :dates, :price_per_night, :reservations, :room_total, :rooms
+    
+    attr_accessor :daterange_factory, :room_factory
+    
+    def initialize(id:, room_numbers:, start_date: nil, end_date: nil, price_per_night:, room_source: nil)
+      @id = id
+      @price_per_night = price_per_night
+      @room_total = room_numbers.length
+      
+      @reservations = []
+      @daterange_factory = DateRangeFactory.new()
+      @room_factory = RoomFactory.new()
+      
+      if !(start_date.nil? && end_date.nil?)
```

Comment:
I think you want this to be:
```suggestion
      if !start_date.nil? && !end_date.nil?
```

Also, you probably want to `raise` an `ArgumentError` here instead of not having dates.